### PR TITLE
Part 5: Officially splitting FENDA and PerFCL clients

### PR DESCRIPTION
**Pull request by emersodb**
# PR Type
Feature

# Short Description

Clickup Ticket(s): [Link](https://app.clickup.com/t/8687xebzz)

This PR is Part 4 of several PRs aiming to simplify and decouple our implementations of FENDA and PerFCL. In this fourth part, we're finally splitting the FENDA and PerFCL clients. As part of this split, we unify the model bases with a "parallel split model base." 

As part of this we add a new PerFCL example.

**NOTE**: In the next PR, we will re-introduce the additional loss components option for FENDA as a separate client called `ConstrainedFendaClient`.

# Tests Added

Migrating a bunch of tests to the new structure. Adding tests explicitly for the PerFCL client. Expanding tests for the FENDA client and other components.

### CODE REVIEW

**Review by sanaAyrml — APPROVED** [review]
Thanks for this PR.

File: fl4health/clients/perfcl_client.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+        device: torch.device,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        checkpointer: Optional[ClientCheckpointModule] = None,
+        global_feature_loss_temperature: float = 0.5,
+        local_feature_loss_temperature: float = 0.5,
+        global_feature_contrastive_loss_weight: float = 1.0,
+        local_feature_contrastive_loss_weight: float = 1.0,
+    ) -> None:
+        """
+        This client is used to perform training client-side training associated with the PerFCL method derived in
```
Comment (sanaAyrml) [inline fl4health/clients/perfcl_client.py:31]: training is extra here

Comment (emersodb) [reply fl4health/clients/perfcl_client.py:31]: Indeed it is!

File: fl4health/clients/fenda_client.py [hunk 2]
```diff
@@ -33,246 +29,23 @@ def __init__(
             loss_meter_type=loss_meter_type,
             checkpointer=checkpointer,
         )
-        """This module is used to init FENDA client with various auxiliary loss functions.
-        These losses will be activated only when their weights are not 0.0.
+        """
+        This client is used to perform training client-side training associated with the FENDA method described in
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/clients/fenda_client.py:33]: See, I also make grammar mistakes 🙂. Will fix.

File: fl4health/model_bases/fenda_base.py [hunk 3]
```diff
@@ -1,58 +1,43 @@
...
-class FendaModel(PartialLayerExchangeModel):
-    def __init__(self, local_module: nn.Module, global_module: nn.Module, model_head: FendaHeadModule) -> None:
-        super().__init__()
-        self.local_module = local_module
-        self.global_module = global_module
-        self.model_head = model_head
 
-    def layers_to_exchange(self) -> List[str]:
-        return [layer_name for layer_name in self.state_dict().keys() if layer_name.startswith("global_module.")]
+class FendaModelWithFeatureState(FendaModel):
```
Comment (jewelltaylor) [inline fl4health/model_bases/fenda_base.py:20]: I will leave it up to your judgement because I don't want you to make documentation for the sake of making documentation but I think for at least some of the model bases you added it mat be helpful to add some documentation to help people figure out what it is and what context it should be used in

File: fl4health/model_bases/perfcl_base.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+    def layers_to_exchange(self) -> List[str]:
+        return [layer_name for layer_name in self.state_dict().keys() if layer_name.startswith("second_module.")]
+
+    def forward(self, input: torch.Tensor) -> Tuple[Dict[str, torch.Tensor], Dict[str, torch.Tensor]]:
+        # input is expected to be of shape (batch_size, *)
+        local_output = self.first_module.forward(input)
+        global_output = self.second_module.forward(input)
+        preds = {"prediction": self.model_head.forward(local_output, global_output)}
+        # PerFCL models always store features from the
```
Comment (jewelltaylor) [inline fl4health/model_bases/perfcl_base.py:24]: nit: PerFCL models always store feature from the extractors (or whatever we want to call them)

File: fl4health/clients/perfcl_client.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+                - The tensor for the total loss
+                - A dictionary with `loss`, `total_loss`, `global_feature_contrastive_loss`, and
+                    `local_feature_contrastive_loss` representing the various and relevant pieces of the loss
+                    calculations
+        """
+
+        loss = self.criterion(preds["prediction"], target)
+        # If any of these are None then we don't compute the PerFCL loss. This will happen on the first client-side
+        # training run.
+        if self.old_local_module is None or self.old_global_module is None or self.initial_global_module is None:
```
Comment (jewelltaylor) [inline fl4health/clients/perfcl_client.py:210]: Should we also assert its the first round and throw something if they are not provided and its not the first round. Or we can log something. Just to raise if we are not computing some of the losses and its not the first round meaning there is likely an error happening somewhere. If you think its unlikely to happen though you can disregard this

File: fl4health/model_bases/fenda_base.py [hunk 6]
```diff
@@ -1,58 +1,43 @@
...
         # input is expected to be of shape (batch_size, *)
-        local_output = self.local_module.forward(input)
-        global_output = self.global_module.forward(input)
+        local_output = self.first_module.forward(input)
+        global_output = self.second_module.forward(input)
         preds = {"prediction": self.model_head.forward(local_output, global_output)}
         features = {
-            "local_features": local_output.reshape(len(local_output), -1),
-            "global_features": global_output.reshape(len(global_output), -1),
+            "local_features": local_output.reshape(len(local_output), -1) if self.flatten_features else local_output,
```
Comment (jewelltaylor) [inline fl4health/model_bases/fenda_base.py:37]: This is a bit hard to read, I suggest you define the dict and then add each key on subsequent lines. Or some other way of breaking this up

**Review by jewelltaylor — COMMENTED** [review]
All in all, looks great! I love that you have decoupled the implementations (so its easier for users without intimate knowledge of the underlying methods) but still have some reuse in the underlying model bases. I have done an initial pass and only had a few comments. I will take another pass through once you have a chance to address these minor comments just to be thorough but I don't imagine anything will come up

File: fl4health/model_bases/fenda_base.py (hunk 3 above)
Comment (emersodb) [reply fl4health/model_bases/fenda_base.py:20]: Yeah I think that's a good call. Will add some here.

File: fl4health/model_bases/parallel_split_models.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            if self.mode == ParallelFeatureJoinMode.CONCATENATE
+            else torch.add(first_tensor, second_tensor)
+        )
+        return self.head_forward(head_input)
+
+
+class ParallelSplitModel(nn.Module):
+    def __init__(self, first_module: nn.Module, second_module: nn.Module, model_head: ParallelSplitHeadModule) -> None:
+        super().__init__()
+        self.first_module = first_module
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/model_bases/parallel_split_models.py:39]: Yeah, I don't love that they aren't called local/global here but they are more general, and so thought the names needed to be. I like your suggestion here though! I'll change the names to those.

File: fl4health/clients/perfcl_client.py (hunk 5 above)
Comment (emersodb) [reply fl4health/clients/perfcl_client.py:210]: Talk about this together, this is a little more complicated to do but we should do it. So a ticket has been filed in the backlog: https://app.clickup.com/t/8688ptp72

File: fl4health/model_bases/fenda_base.py (hunk 6 above)
Comment (emersodb) [reply fl4health/model_bases/fenda_base.py:37]: Will take a shot at making it a bit more readable 🙂

**Review by jewelltaylor — APPROVED** [review]
Added documentation looks great!
